presto-cassandra/src/main/java/com/facebook/presto/cassandra/NativeCassandraSession.java

```diff
@@ -420,6 +428,88 @@ private CassandraColumnHandle buildColumnHandle(AbstractTableMetadata tableMetad
+        }
+
+        ByteBuffer buffer = ByteBuffer.allocate(1000);
+        HashMap<ColumnHandle, NullableValue> map = new HashMap<>();
+        Set<String> uniquePartitionIds = new HashSet<>();
+        StringBuilder stringBuilder = new StringBuilder();
+
+        boolean isComposite = partitionKeyColumns.size() > 1;
+
+        ImmutableList.Builder<CassandraPartition> partitions = ImmutableList.builder();
+        for (List<Object> values : Sets.cartesianProduct(filterPrefixes)) {
+            buffer.clear();
+            map.clear();
+            stringBuilder.setLength(0);
+            for (int i = 0; i < partitionKeyColumns.size(); i++) {
+                Object value = values.get(i);
+                CassandraColumnHandle columnHandle = partitionKeyColumns.get(i);
+                CassandraType cassandraType = columnHandle.getCassandraType();
+
+                switch (cassandraType) {
+                    case TEXT:
+                        Slice slice = (Slice) value;
+                        if (isComposite) {
+                            buffer.putShort((short) slice.length());
+                            buffer.put(slice.getBytes());
+                            buffer.put((byte) 0);
+                        }
+                        else {
+                            buffer.put(slice.getBytes());
+                        }
+                        break;
+                    case INT:
+                        int intValue = toIntExact((long) value);
+                        if (isComposite) {
+                            buffer.putShort((short) Integer.BYTES);
+                            buffer.putInt(intValue);
+                            buffer.put((byte) 0);
+                        }
+                        else {
+                            buffer.putInt(intValue);
+                        }
+                        break;
+                    case BIGINT:
+                        if (isComposite) {
+                            buffer.putShort((short) Long.BYTES);
+                            buffer.putLong((long) value);
+                            buffer.put((byte) 0);
+                        }
+                        else {
+                            buffer.putLong((long) value);
+                        }
+                        break;
+                    default:
+                        throw new IllegalStateException("Handling of type " + cassandraType + " is not implemented");
+                }
+
+                map.put(columnHandle, NullableValue.of(cassandraType.getNativeType(), value));
+                if (i > 0) {
+                    stringBuilder.append(" AND ");
+                }
+                stringBuilder.append(CassandraCqlUtils.validColumnName(columnHandle.getName()));
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraType.getColumnValueForCql(value, cassandraType));
+            }
+            buffer.flip();
```

Comment:
Why the buffer does `flip`?
I guess `CassandraPartition` requirements. but I want to know exact reason.

Reply:
I convert ByteBuffer to byte array to create a `CassandraPartition` in the same way of `NativeCassandraSession#getPartitions`.  As I mention in https://github.com/reproio/presto/pull/1/files#r229854855, we should build partition keys in the same way as Cassandra does (cf. [CFMetaData#serializePartitionKey](https://github.com/apache/cassandra/blob/cassandra-3.11.3/src/java/org/apache/cassandra/config/CFMetaData.java#L793-L806)).

```
java> ByteBuffer buf = ByteBuffer.allocate(10)
java.nio.ByteBuffer buf = java.nio.HeapByteBuffer[pos=0 lim=10 cap=10]
java> buf.putInt(1)
java.nio.ByteBuffer res6 = java.nio.HeapByteBuffer[pos=4 lim=10 cap=10]
java> buf.array()
byte[] res7 = [0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
java> buf.limit()
java.lang.Integer res8 = 10
java> buf.flip()
java.nio.ByteBuffer res9 = java.nio.HeapByteBuffer[pos=0 lim=4 cap=10]
java> byte[] key = new byte[buf.limit()]
byte[] key = [0, 0, 0, 0]
java> buf.get(key)
java.nio.ByteBuffer res11 = java.nio.HeapByteBuffer[pos=4 lim=4 cap=10]
java> key
byte[] key = [0, 0, 0, 1]
```

Reply:
I understand. Thanks!!